agent/agent.go

```diff
@@ -94,6 +100,86 @@ type outputUnit struct {
 	outputs []*models.RunningOutput
 }
 
+func (a *Agent) runSingleInput(input_config *models.InputConfig, ctx context.Context) {
+	startTime := time.Now()
+
+	plugin := inputs.Inputs[input_config.Name]
+	input := models.NewRunningInput(plugin(), input_config)
+
+	// Overwrite agent interval if this plugin has its own.
+	interval := a.Config.Agent.Interval.Duration
+	if input.Config.Interval != 0 {
+		interval = input.Config.Interval
+	}
+
+	// Overwrite agent precision if this plugin has its own.
+	precision := a.Config.Agent.Precision.Duration
+	if input.Config.Precision != 0 {
+		precision = input.Config.Precision
+	}
+
+	// Overwrite agent collection_jitter if this plugin has its own.
+	jitter := a.Config.Agent.CollectionJitter.Duration
+	if input.Config.CollectionJitter != 0 {
+		jitter = input.Config.CollectionJitter
+	}
+
+	var ticker Ticker
+	if a.Config.Agent.RoundInterval {
+		ticker = NewAlignedTicker(startTime, interval, jitter)
+	} else {
+		ticker = NewUnalignedTicker(interval, jitter)
+	}
+
+	acc := NewAccumulator(input, a.iu.dst)
+	acc.SetPrecision(getPrecision(precision, interval))
+
+	a.wg.Add(1)
+	go func(input *models.RunningInput) {
+		defer ticker.Stop()
+		defer a.wg.Done()
+		a.gatherLoop(ctx, acc, input, ticker, interval)
+	}(input)
+
+	a.Config.Inputs = append(a.Config.Inputs, input)
+}
+
+func (a *Agent) AddInputPluginToAgent(pluginName string) {
+	plugin := inputs.Inputs[pluginName]
+	inputConfig := models.InputConfig{
+		Name: pluginName,
+		// TODO rest of config?
+	}
+	runningPlugin := models.NewRunningInput(plugin(), &inputConfig)
+	a.Config.Inputs = append(a.Config.Inputs, runningPlugin)
+}
+
+func (a *Agent) AddOutputPluginToAgent(pluginName string) {
+	plugin := outputs.Outputs[pluginName]
```

Comment:
error check? this will panic and terminate the entire process if someone puts in an invalid pluginName when it calls ```plugin()```. Plus putting in error stuff now will be good for future when needing to validate the new config as well as making sure things are written back to disk